### ISSUE TITLE
Could not fetch authorities at ...: InvalidAuthoritiesSet

### ISSUE THREAD

**Issue by kostekIV** [paritytech/substrate#10453]
So we have encountered this error and we are not sure why this happens (if this is something that should be handled on our side or somewhere in the substrate).

The setup that this error appears is as follows:
* We are using Aura for block proposals.
* For some height `h` we have two blocks `A` and `B`.
* On top of the block `A`, we already have imported `C`
* We are currently importing block `D` which builds on top of `C`
* In the meantime, block `B` is finalized.
* The import of `D` fails with the mentioned error.

Not all of the nodes fail with the import tho. Some of them successfully reorganize onto a chain that builds on the `B` block after `B` is finalized.

Because of this error the finalization stales, here again, we are not sure yet whether this is something we should handle in our finalization gadget that we are not aware of, or maybe we have some dirty state (resetting nodes helps but is not an ideal solution.)

Thanks

**Comment by DamianStraszak** [comment]
I'm working with @kostekIV and today we have realized what might be a possible explanation for this.

First of all, here is a visualization of the above situation:
![Screenshot from 2021-12-10 17-06-40](https://user-images.githubusercontent.com/23038619/145604676-f1299aa6-f3a6-4097-9bf3-9479f91be27b.png)

In the depicted situation the import of block D seems to fail, with the cryptic error 
```
Verification failed for block 0x7467e75d032920abda959c28c89ed93c863c304a7000e11483740e14bae09585 received from peer: 12D3KooWD5s2dkifJua69RbLwEREDdJjsNHvavNRGxdCvzhoeaLc, "Could not fetch authorities at 0xcd92a6d2a064ed96dbe199d5f3ec4d65d4fa65e5c4c1b9e79176d8adb1d826af: InvalidAuthoritiesSet"
```

Now, the point is that in our custom finality gadget (replacement for grandpa) we certainly MUST import this block, and we even request it explicitly via `NetworkService::set_sync_fork_request` -- it still fails.

The diagnosis we currently have, is that there must be some state pruning happening for blocks outside of the finalized branch and that is the reason. Could this be correct? 

In other words: suppose we have a block `D`  that is a fork such that a conflicting block has been already finalized. Is it the expected behavior that **importing such a block should/might fail**?

**Comment by bkchr** [comment]
CC @andresilva

**Comment by DamianStraszak** [comment]
Any hints on that?

**Comment by andresilva** [comment]
> In other words: suppose we have a block D that is a fork such that a conflicting block has been already finalized. Is it the expected behavior that importing such a block should/might fail?

If you are running with archive mode for state (default mode for validators, no state pruning) then it should be possible to import the block. Otherwise, if pruning is enabled we will automatically prune all states that are not in the finalized chain. I believe we should fallback to importing just the header (since that doesn't require state as the block is not executed), but would have to double check that.

**Comment by DamianStraszak** [comment]
Thanks for the answer @andresilva!  We run with pruning **disabled**, but there are still issues with importing such blocks...

Is there then any special logic regarding importing such blocks? (=blocks which branch out of the finalized chain, thus cannot be ever finalized) Or any special treatment for such blocks, or their state?

**Comment by andresilva** [comment]
Could you try collecting some logs of this happening with `-lsync=trace`? Also, in the example above can you tell whether the remaining blocks of the stale fork were imported and only `D` failed? cc @arkpar

**Comment by arkpar** [comment]
>We are currently importing block D which builds on top of C
In the meantime, block B is finalized.

This should not be happening in parallel. Import and finalisation events are serialized. Verification however indeed happens before the import so it my produce such errors. As soon as `B` is finalized `A` is discarded unless you run in archive mode. Pruning history limit only sets how many blocks to keep in the **canonical** chain.

**Comment by DamianStraszak** [comment]
Thank you both @andresilva and @arkpar for responses. 

> Could you try collecting some logs of this happening with `-lsync=trace`? Also, in the example above can you tell whether the remaining blocks of the stale fork were imported and only `D` failed? 

Yes, I will deliver such logs, just need some time. In this example, `A` and `C` were imported, and apart from that there were thousands of attempts to import `D`, all with the same error. I don't know if there existed any other blocks beyond `D` in this branch, but certainly they were not imported by this node.

We run with the `--validator` flag (and no other pruning-related flags), which AFAIK disables pruning, correct?

**Comment by DamianStraszak** [comment]
I have been trying to reproduce the issue and while I could occasionally trigger the `InvalidAuthoritiesSet` error, it was not so frequent and persistent as before... However, I've encountered numerous 

```Err(Other(ClientImport("Potential long-range attack: block not in finalized chain.")))```
errors. I'm attaching logs -- [long-range.txt](https://github.com/paritytech/substrate/files/7802160/long-range.txt) -- the interesting block is `0x82ecce9196567ef95ef282f31c0832121274c1fd8cd79b591f01e0940e090dd8` 


Now I'm thinking that the `InvalidAuthoritiesSet` is somehow connected to that and while I don't know the exact reason it shows up, it has certainly something to do with importing "hopeless forks" (blocks that can never possibly be finalized). In particular, after inspecting the code in `parity/substrate/client/db/src/lib.rs` and few other files I can answer my previous question.

> Is there then any special logic regarding importing such blocks? (=blocks which branch out of the finalized chain, thus cannot be ever finalized) Or any special treatment for such blocks, or their state?

There is special logic for that -- at several places the `NotInFinalizedChain` exception is thrown in such a case. Which essentially makes it impossible to import such blocks upon request. Is that correct?

**Comment by andresilva** [comment]
You are correct, I think https://github.com/paritytech/substrate/blob/master/client/service/src/client/client.rs#L541-L548 is preventing the import of any forked blocks that are behind the latest finalized block. I believe we should still be able to import those blocks but making sure that we don't change the chain head regardless of the block number of the block being imported. @arkpar do you see any reason for not doing this?

**Comment by arkpar** [comment]
I guess the reason is not to waste computation and storage on such blocks. What would be the reason to include them? Uncle rewards?

**Comment by DamianStraszak** [comment]
Now that we learned that this is the intended behavior, we can actually make some changes on our side to circumvent this issue. So, the only question that remains is whether attempts of importing such blocks can cause the `InvalidAuthoritiesSet` error.

Let me also offer a few words of explanation of why is this a problem for us. We use our custom finality-gadget AlephBFT (in place of GRANDPA) and not being able to import such blocks is problematic for our data-availability mechanism. More specifically, the consensus messages in AlephBFT carry block hashes, and whenever such a message is received, we make efforts to fetch the appropriate blocks (because finalizing hashes for which we don't know the underlying data is obviously unsafe). Now, we run into troubles when we get a hash `h`, we don't know the underlying block, and it corresponds to a block like `D` in the picture above (imagine we didn't import `C` either). In such a case we have no idea what `D` is and whether we should attempt fetching it, or conclude it is irrelevant because it will never be finalized. 

That being said, we have already found a fix for that on our side, and the most important thing was to get some confirmation of our understanding on how things work from you guys! Thanks for all the responses!
